### PR TITLE
test/e2e/network: fix a bug in the hostport e2e test

### PR DESCRIPTION
**Pull request by aspsk**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The hostport e2e test (sonobuoy run --e2e-focus 'validates that there is no
conflict between pods with same hostPort but different hostIP and protocol')
checks, in particular, that two pods with the same hostPort, the same hostIP,
but different L4 protocols can coexist on one node.

In order to do this, the test creates two pods with the same hostIP:hostPort,
one TCP-based, another UDP-based. However, both pods listen on both protocols:

    netexec --http-port=8080 --udp-port=8080

This can happen that a CNI which doesn't distinguish between TCP and UDP
hostPorts forwards all traffic, TCP or UDP, to the same pod. As this pod
listens on both protocols it will reply to both requests, and the test
will think that everything works properly while the second pod is indeed
disconnected. Fix this by executing different commands in different pods:

    TCP: netexec --http-port=8080 --udp-port=-1
    UDP: netexec --http-port=8008 --udp-port=8080

The TCP pod now doesn't listen on UDP, and the UDP pod doesn't listen on TCP on
the target hostPort. The UDP pod still needs to listen on TCP on another port
so that a pod readiness check can be made.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Review by aanm — APPROVED** [review]

**Comment by aojea** [comment]
/ok-to-test
good catch, but should we add a test to check that only one protocol is forwarded?
If I open a TCP port and the traffic in UDP is forwarded too, it sounds like a bug to me

**Review by borkmann — APPROVED** [review]

**Comment by aspsk** [comment]
> should we add a test to check that only one protocol is forwarded?
> If I open a TCP port and the traffic in UDP is forwarded too, it sounds like a bug to me

Yes, sounds like a good idea. I can do it in a later PR (don't have cycles now, but I will return to this task later anyway)

File: test/e2e/network/hostport.go [hunk 1]
```diff
@@ -157,6 +157,18 @@ var _ = common.SIGDescribe("HostPort", func() {
...
+	var netexecArgs []string
+	var readinessProbePort int32
+
+	if protocol == v1.ProtocolTCP {
+		readinessProbePort = 8080
+		netexecArgs = []string{"--http-port=8080", "--udp-port=-1"}
+	} else {
+		readinessProbePort = 8008
+		netexecArgs = []string{"--http-port=8008", "--udp-port=8080"}
+	}
```
Comment (aojea) [inline test/e2e/network/hostport.go:170]: we support SCTP too, the test only checks UDP though but do you mind fixing this function just for all protocols

Comment (aojea) [reply test/e2e/network/hostport.go:170]: https://github.com/kubernetes/kubernetes/blob/1290d6a4c0c5cc4f758b428993d2c3760d9cf279/test/images/agnhost/netexec/netexec.go#L140-L141

**Comment by aojea** [comment]
shouldn't the test fail here in this situation?

https://github.com/kubernetes/kubernetes/blob/8c76845b0389e67c5d2d51343d4122d110d279aa/test/e2e/network/hostport.go#L139-L143

we are checking that the answer doesn't come from the same pod

**Comment by aojea** [comment]
> we are checking that the answer doesn't come from the same pod

nevermind, that is not for the UDP

**Comment by aspsk** [comment]
> > we are checking that the answer doesn't come from the same pod
> 
> nevermind, that is not for the UDP

Actually, you're right. For the TCP vs. UDP case we can just get `hostname3` in the third exec command, and check that `hostname2 != hostname3`

**Comment by aojea** [comment]
this will be the best solution 
```diff
diff --git a/test/e2e/network/hostport.go b/test/e2e/network/hostport.go
index dec0d359949..436d7e081e7 100644
--- a/test/e2e/network/hostport.go
+++ b/test/e2e/network/hostport.go
@@ -119,7 +119,7 @@ var _ = common.SIGDescribe("HostPort", func() {
                // IPv6 doesn't NAT from localhost -> localhost, it doesn't have the route_localnet kernel hack, so we need to specify the source IP
                cmdPod1 := []string{"/bin/sh", "-c", fmt.Sprintf("curl -g --connect-timeout %v --interface %s http://%s/hostname", timeout, hostIP, net.JoinHostPort(localhost, strconv.Itoa(int(port))))}
                cmdPod2 := []string{"/bin/sh", "-c", fmt.Sprintf("curl -g --connect-timeout %v http://%s/hostname", timeout, net.JoinHostPort(hostIP, strconv.Itoa(int(port))))}
-               cmdPod3 := []string{"/bin/sh", "-c", fmt.Sprintf("nc -vuz -w %v %s %d", timeout, hostIP, port)}
+               cmdPod3 := []string{"/bin/sh", "-c", fmt.Sprintf("echo hostname | nc -u -w %v %s %d", timeout, hostIP, port)}
                // try 5 times to connect to the exposed ports
                for i := 0; i < 5; i++ {
                        // check pod1
@@ -143,15 +143,29 @@ var _ = common.SIGDescribe("HostPort", func() {
                        }
                        // check pod3
                        ginkgo.By(fmt.Sprintf("checking connectivity from pod %s to serverIP: %s, port: %d UDP", hostExecPod.Name, hostIP, port))
-                       _, _, err = f.ExecCommandInContainerWithFullOutput(hostExecPod.Name, "e2e-host-exec", cmdPod3...)
+                       hostname3, _, err = f.ExecCommandInContainerWithFullOutput(hostExecPod.Name, "e2e-host-exec", cmdPod3...)
                        if err != nil {
                                framework.Logf("Can not connect from %s to pod(pod2) to serverIP: %s, port: %d", hostExecPod.Name, hostIP, port)
                                continue
                        }
+                       if hostname1 == hostname3 {
+                               framework.Logf("pods must have different hostname: pod1 has hostname %s, pod3 has hostname %s", hostname1, hostname3)
+                               continue
+                       }
+                       if hostname2 == hostname3 {
+                               framework.Logf("pods must have different hostname: pod2 has hostname %s, pod3 has hostname %s", hostname2, hostname3)
+                               continue
+                       }
+                       // success
                        return
                }
                framework.Failf("Failed to connect to exposed host ports")
        })
+
+       ginkgo.It("validates that only the specified protocol is forwarded [LinuxOnly]", func() {
+
+       })
+
 })
```

File: test/e2e/network/hostport.go (hunk 1 above)
Comment (aojea) [reply test/e2e/network/hostport.go:170]: scratch this comment

**Comment by aojea** [comment]
this is a conformance test, can we better use this approach https://github.com/kubernetes/kubernetes/pull/110640#issuecomment-1158938674? it seems more correct

**Comment by aspsk** [comment]
> this is a conformance test, can we better use this approach [#110640 (comment)](https://github.com/kubernetes/kubernetes/pull/110640#issuecomment-1158938674)? it seems more correct

Yes, thanks, your patch looks correct. I will test it locally and update the commit

**Comment by aojea** [comment]
you have a format error 
```
{Script Error ScriptError diff -u ./test/e2e/network/hostport.go.orig ./test/e2e/network/hostport.go
--- ./test/e2e/network/hostport.go.orig	2022-06-17 16:38:53.420795817 +0000
+++ ./test/e2e/network/hostport.go	2022-06-17 16:38:53.420795817 +0000
@@ -150,11 +150,11 @@
 			}
 			if hostname1 == hostname3 {
 				framework.Logf("pods must have different hostname: pod1 has hostname %s, pod3 has hostname %s", hostname1, hostname3)
-					continue
+				continue
 			}
 			if hostname2 == hostname3 {
 				framework.Logf("pods must have different hostname: pod2 has hostname %s, pod3 has hostname %s", hostname2, hostname3)
-					continue
+				continue
 			}
 			return
 		}

Run ./hack/update-gofmt.sh}
```

otherwise it sgtm

File: test/e2e/network/hostport.go [hunk 2]
```diff
@@ -143,11 +143,19 @@ var _ = common.SIGDescribe("HostPort", func() {
...
 			ginkgo.By(fmt.Sprintf("checking connectivity from pod %s to serverIP: %s, port: %d UDP", hostExecPod.Name, hostIP, port))
-			_, _, err = f.ExecCommandInContainerWithFullOutput(hostExecPod.Name, "e2e-host-exec", cmdPod3...)
+			hostname3, _, err := f.ExecCommandInContainerWithFullOutput(hostExecPod.Name, "e2e-host-exec", cmdPod3...)
 			if err != nil {
 				framework.Logf("Can not connect from %s to pod(pod2) to serverIP: %s, port: %d", hostExecPod.Name, hostIP, port)
 				continue
 			}
+			if hostname1 == hostname3 {
+				framework.Logf("pods must have different hostname: pod1 has hostname %s, pod3 has hostname %s", hostname1, hostname3)
+					continue
```
Comment (aojea) [inline test/e2e/network/hostport.go:153]: gofmt complains because of the wrong indentation

File: test/e2e/network/hostport.go [hunk 3]
```diff
@@ -143,11 +143,19 @@ var _ = common.SIGDescribe("HostPort", func() {
...
+			}
+			if hostname2 == hostname3 {
+				framework.Logf("pods must have different hostname: pod2 has hostname %s, pod3 has hostname %s", hostname2, hostname3)
+					continue
```
Comment (aojea) [inline test/e2e/network/hostport.go:157]: ditto

Comment (aspsk) [reply test/e2e/network/hostport.go]: Yeah, I wonder how this happened, because I did test this locally. Hmm.

I've pushed a new version

Comment (aspsk) [reply test/e2e/network/hostport.go]: Ok, I think I know what happened. Before pushing I did press `=` in visual mode in vim. This is, probably, what I do subconsciously when writing in C :rofl: Need to get used to Go

Comment (aojea) [reply test/e2e/network/hostport.go]: :smile:

**Comment by aojea** [comment]
/lgtm
/approve

Thanks

**Comment by hswong3i** [comment]
When testing with kubernetes v1.25.0-rc.0 (see https://github.com/cncf/k8s-conformance/pull/2093), I also get a fail with `[sig-network] HostPort validates that there is no conflict between pods with same hostPort but different hostIP and protocol [LinuxOnly] [Conformance]`.

BTW, corresponding test passed with kubernetes v1.24.3 (see https://github.com/cncf/k8s-conformance/commit/2a737c5c7d205982a8e710b1dd1b39ace4c47e6c).

I found that this PR already get merged into v1.25.0-rc.0 so should that suppose to be fixed? I am testing both with sonobuoy 0.56.9 so am I missing something?

**Comment by hswong3i** [comment]
Re-test with Ubuntu 22.04 but still failed: https://github.com/cncf/k8s-conformance/pull/2093/commits/9e75f135a8863dc527225bba976fbe0a0a3c15c5

**Pull request merged**